### ISSUE TITLE
[Bug] Wrong 'selected card count' in card browser when duplicate cards exist.

### ISSUE THREAD

**Issue by sumadithya** [ankidroid/Anki-Android#8547]
Original error message below. It was useful but further investigation shows that it's not the root cause.

# Reproduction step

1. Open card browser 
2. search for plenty of cards (on my table, more than 61)
3. Select first one
4. Scroll down

# Expected result

No other card is selected

# Actual result

Another card is selected, which is a duplicate of the first one. The fact that they are duplicate is the actual error

# Original error message

###### Reproduction Steps

1. Download a [deck which contains duplicate cards](https://ankiweb.net/shared/info/1225334140)
2. Go to card browser.
3. Select the the deck which is mentioned above.
4. It shows '549' as total count.
5. Select All
6. It shows '508' as count selected.

###### Expected Result
* Count for selected cards and the total number of cards should be equal when the option 'select all' is selected.

###### Actual Result
* Selected count is less than the total count when using 'select all'

###### Debug info
AnkiDroid Version = 2.15alpha38

Android Version = 9

Manufacturer = Xiaomi

Model = Redmi 6 Pro

Hardware = qcom

Webview User Agent = Mozilla/5.0 (Linux; Android 9; Redmi 6 Pro Build/PKQ1.180917.001; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/89.0.4389.105 Mobile Safari/537.36

ACRA UUID = 9797d9fd-54ac-438b-a012-a69a5cdc3c6d

Scheduler = std

Crash Reports Enabled = false

DatabaseV2 Enabled = true

###### Research
- [x] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [x] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [x] I have searched for similar existing issues here and on the user forum
- [ ] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

**Comment by sumadithya** [comment]
You need not select all cards to replicate this.
If a card has a duplicate, select it, its copy is selected automatically but the count that displays on top doesn't account for duplicates.

**Comment by sumadithya** [comment]
![image](https://user-images.githubusercontent.com/27889757/114272621-7feac000-9a34-11eb-8af8-89de20bd2595.png)

You can't see it clearly in this image but take my word for it. This is the top of the list of card of the deck mentioned (this is the only thing you'd have to take my word for). I selected two cards in the image. But its actually four cards, since the duplicates are automatically selected. To prove the existence of duplicates, I uploaded the image below.

![image](https://user-images.githubusercontent.com/27889757/114272609-77928500-9a34-11eb-8fb6-f02363ffec15.png)

**Comment by sumadithya** [comment]
I tested it further, pressing and holding an already existing duplicate crashes the app. (I will mention this in another issue)

**Comment by dorrin-sot** [comment]
Hi Is this fixed? Can I work on this

**Comment by david-allison-1** [comment]
@dorrin-sot Thanks for the enthusiasm!

This is blocking 2.15 from being released, I'll take it on. The crash is a separate issue.

Feel free to pick up one of our "Good First Issues" or "Help Wanted" issues

Best,

David

**Comment by david-allison-1** [comment]
Seems to be caused by 3e7983670370141b6dc2e1ab3d7a4036c0585def 

@Arthur-Milchior 

Regression test:

```java
/*
 *  Copyright (c) 2021 David Allison <davidallisongithub@gmail.com>
 *
 *  This program is free software; you can redistribute it and/or modify it under
 *  the terms of the GNU General Public License as published by the Free Software
 *  Foundation; either version 3 of the License, or (at your option) any later
 *  version.
 *
 *  This program is distributed in the hope that it will be useful, but WITHOUT ANY
 *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 *  PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along with
 *  this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.ichi2.async;

import com.ichi2.anki.CardBrowser;
import com.ichi2.anki.RobolectricTest;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;

import java.util.List;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

@RunWith(AndroidJUnit4.class)
public class CollectionTaskSearchCardsTest extends RobolectricTest {

    @Test
    public void test() {
        addNoteUsingBasicModel("Hello", "World");
        addNoteUsingBasicModel("One", "Two");

        TaskListener<List<CardBrowser.CardCache>, List<CardBrowser.CardCache>> listener = mock(TaskListener.class);

        TaskManager.launchCollectionTask(new CollectionTask.SearchCards("", false, 1, 0, 0), listener);

        waitForAsyncTasksToComplete();


        verify(listener, times(1)).onPreExecute();

        ArgumentCaptor<List> argumentCaptor = ArgumentCaptor.forClass(List.class);
        verify(listener, times(1)).onProgressUpdate(argumentCaptor.capture());
        assertThat(argumentCaptor.getValue().size(), is(1)); // 4

        ArgumentCaptor<List> argumentCaptor2 = ArgumentCaptor.forClass(List.class);
        verify(listener, times(1)).onPostExecute(argumentCaptor2.capture());
        assertThat(argumentCaptor.getValue().size(), is(2)); // 4

    }
}
```

**Comment by Arthur-Milchior** [comment]
Is there any technics you use to do a manual bisect in cases like this where the code do not work on old commit?
Because I tried to run it on the commit before 3e79836 and it obviously fails.

I did re read the part of the code in this commit around Searchcards and I don't see any obvious change. But I've no idea what argument captor and mock does

**Comment by david-allison-1** [comment]
> Is there any technics you use to do a manual bisect in cases like this where the code do not work on old commit?
> Because I tried to run it on the commit before [3e79836](https://github.com/ankidroid/Anki-Android/commit/3e7983670370141b6dc2e1ab3d7a4036c0585def) and it obviously fails.
> 
> I did re read the part of the code in this commit around Searchcards and I don't see any obvious change. But I've no idea what argument captor and mock does

I tested via building and running the app, as I couldn't rebase in a unit test due to the structure of the code changing

**Comment by Arthur-Milchior** [comment]
To be clear, the problem is not that the deck as duplicate card. It's that the same card appear twice in the browser. I edit the top message to indicate it

**Comment by david-allison-1** [comment]
I'll have time to work on this today

**Comment by Joseph9Morgan** [comment]
>  dupe_crash.mp4 
> I tested it further, pressing and holding an already existing duplicate crashes the app. (I will mention this in another issue)

Hi ! Could you further elaborate on the crash information? I'm trying to reproduce the crash but I can't find another issue with detailed debug  information.

**Comment by david-allison-1** [comment]
> > dupe_crash.mp4
> > I tested it further, pressing and holding an already existing duplicate crashes the app. (I will mention this in another issue)
> 
> Hi ! Could you further elaborate on the crash information? I'm trying to reproduce the crash but I can't find another issue with detailed debug information.

Hi, I believe this was fixed in https://github.com/ankidroid/Anki-Android/pull/8642. Is there a reason you feel this is still occurring?

**Comment by Joseph9Morgan** [comment]
> > > dupe_crash.mp4
> > > I tested it further, pressing and holding an already existing duplicate crashes the app. (I will mention this in another issue)
> > 
> > 
> > Hi ! Could you further elaborate on the crash information? I'm trying to reproduce the crash but I can't find another issue with detailed debug information.
> 
> Hi, I believe this was fixed in #8642. Is there a reason you feel this is still occurring?

Sorry to have bothered. It’s not occurring in the newest release, but I’m trying to manually reproduce the crash process that aforementioned in the video.
